includes/subscriptions/class-wc-payments-product-service.php

```diff
@@ -115,40 +131,78 @@ public static function get_wcpay_product_hash( WC_Product $product ): string {
+	 * @throws Exception
 	 */
-	public function get_wcpay_product_id( WC_Product $product, $test_mode = null ): string {
+	public function get_or_create_wcpay_product_id( WC_Product $product, $test_mode = null ): string {
 		// If the subscription product doesn't have a WC Pay product ID, create one.
-		if ( ! self::has_wcpay_product_id( $product, $test_mode ) ) {
+		if ( ! $this->has_wcpay_product_id( $product, $test_mode ) ) {
 			$is_current_environment = null === $test_mode || WC_Payments::mode()->is_test() === $test_mode;
 
 			// Only create a new wcpay product if we're trying to fetch a wcpay product ID in the current environment.
 			if ( $is_current_environment ) {
-				WC_Payments_Subscriptions::get_product_service()->create_product( $product );
+				$this->create_product( $product );
 			}
 		}
 
-		return $product->get_meta( self::get_wcpay_product_id_option( $test_mode ), true );
+		return $product->get_meta( self::get_wcpay_product_id_meta_key( $test_mode ), true );
 	}
 
 	/**
-	 * Gets the WC Pay product ID associated with a WC product.
+	 * Get the WCPay product ID for an item type.
 	 *
-	 * @param string $type The item type to create a product for.
-	 * @return string       The item's WCPay product id.
+	 * @param string $type The item type.
+	 *
+	 * @return string The WCPay product ID.
+	 * @throws API_Exception
 	 */
 	public function get_wcpay_product_id_for_item( string $type ): string {
-		$sanitized_type  = self::sanitize_option_key( $type );
-		$option_key_name = self::get_wcpay_product_id_option() . '_' . $sanitized_type;
-		if ( ! get_option( $option_key_name ) ) {
-			$this->create_product_for_item_type( $sanitized_type );
+		$sanitized_type   = self::sanitize_option_key( $type );
+		$option_key_name  = self::get_wcpay_product_id_option() . '_' . $sanitized_type;
+		$wcpay_product_id = get_option( $option_key_name );
+
+		// Case 1: No product found, create a new one.
+		if ( ! $wcpay_product_id ) {
+			return $this->create_product_for_item_type( $sanitized_type );
 		}
-		return get_option( $option_key_name );
+
+		// For existing products, check the linked account.
+		$linked_option_key = $option_key_name . '_linked_to';
+		$linked_account_id = get_option( $linked_option_key );
+		$stripe_account_id = $this->get_stripe_account_id();
+
+		// Case 2: Product exists but linked account doesn't, validate and update if needed.
+		if ( ! $linked_account_id ) {
+			try {
+				// Validate that the product exists for the current account.
+				$existing_product = $this->payments_api_client->get_product_by_id( $wcpay_product_id );
+
+				if ( $existing_product ) {
+					// Product exists, save with current account ID.
+					$this->save_wcpay_product_data( $wcpay_product_id, $stripe_account_id, $sanitized_type );
+					return $wcpay_product_id;
+				} else {
+					// Product doesn't exist, create new one.
+					return $this->create_product_for_item_type( $sanitized_type );
```

Comment:
The case block above suggests that the product exists and the subsequent handling is under that assumption but here we are creating a new product so it seems confusing. I am not sure whether we should create a new product in this scenario or bail out and log that this happened, and then suggesting the end merchant steps to resolve the problem. 🤔

Reply:
Good point, but I think we need to create a new product here. The issue is that the user can have product ID already stored, but not linked, which means that this product was created before this PR. When the merchant installs the plugin with this PR, and before anything else or before any purchase, the merchant re-onboards. Now we have a product ID stored in a legacy way where the stored product is not linked to a specific Stripe account ID, but it belongs to a totally different stripe account id. To be sure, and to avoid linking a product id that belongs to someone else, we fetch the product by ID to confirm that the legacy or stored product ID actually belongs to that Stripe account ID and not to someone else.